src/Runner.Worker/CompositeStepsRunner.cs

```diff
@@ -0,0 +1,249 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using GitHub.DistributedTask.Expressions2;
+using GitHub.DistributedTask.ObjectTemplating.Tokens;
+using GitHub.DistributedTask.Pipelines;
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.Pipelines.ObjectTemplating;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Runner.Common;
+using GitHub.Runner.Common.Util;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker.Expressions;
+using ObjectTemplating = GitHub.DistributedTask.ObjectTemplating;
+using Pipelines = GitHub.DistributedTask.Pipelines;
+
+namespace GitHub.Runner.Worker
+{
+    [ServiceLocator(Default = typeof(CompositeStepsRunner))]
+    public interface ICompositeStepsRunner : IRunnerService
+    {
+        Task RunAsync(IExecutionContext Context);
+    }
+
+
+    public sealed class CompositeStepsRunner : RunnerService, ICompositeStepsRunner
+    {
+        public async Task RunAsync(IExecutionContext actionContext)
+        {
+            // Another approach we can explore, is moving all this logic to the CompositeActionHandler if it's small enough. 
+
+            ArgUtil.NotNull(actionContext, nameof(actionContext));
+            ArgUtil.NotNull(actionContext.CompositeSteps, nameof(actionContext.CompositeSteps));
+
+            // The parent StepsRunner of the whole Composite Action Step handles the cancellation stuff already. 
+            while (actionContext.CompositeSteps.Count > 0)
```

Comment:
I think you can just foreach composite steps since the list won't change while steps are running.
the `while` loop was for handling `post` steps that are registered while steps are running.